### PR TITLE
Robust backward compatibility test

### PR DESCRIPTION
**Pull request by zpoint**
## Whats changed:

* Add another parameter to specify the base branch in the argument instead of using `master`. This allows us to use this parameter for release testing.  
* After `source`, also `cd` into the directory to ensure the `pwd` always points to the running sky code.
* Restart the sky API server whenever venv change and when the script exits.
* Suppress the conda debug logs, as they are much less relevant than our normal logs.

## Description:

Fix the issue where #4683 failed to [catch the backward compatibility problem](https://github.com/skypilot-org/skypilot/pull/4683#discussion_r1955519017).

The issue is because:

The two environments just use `conda activate` to switch the environment but don’t change the directory. This works locally since all imports are from the lib. However, buildkite needs to set `PYTHONPATH=":$(pwd)"` to make the pytest command work, which prioritizes imports from the local path(`pwd`), As a result, both environments use the sky codes from the current path, and the same code remains compatible. 
image
image

And also fix [this](https://github.com/skypilot-org/skypilot/pull/4741#discussion_r1966446347) for client server broken tests.

Tested (run the relevant ones):

- [x] Backward incompatiable: `/quicktest-core` [fail](https://buildkite.com/skypilot-1/quicktest-core/builds/146#_) on buildkite
- [x] Backward compatiable: `/quicktest-core` [success](https://buildkite.com/skypilot-1/quicktest-core/builds/135#_) on buildkite

### CODE REVIEW

File: sky/cli.py [hunk 1]
```diff
@@ -1224,6 +1252,9 @@ def launch(
         # job_id will be None if no job was submitted (e.g. no entrypoint
         # provided)
         if not detach_run and job_id is not None:
+            # The cluster could be in INIT state which cause tail_logs raise
+            # sky.exceptions.ClusterNotUpError
+            _wait_until_cluster_status_not_init(handle.get_cluster_name())
```
Comment (zpoint) [inline sky/cli.py:1257]: Fix the fail found during buildkite [test](https://buildkite.com/skypilot-1/quicktest-core/builds/122#01951821-d1d8-47d4-bd43-31c55b318534)

```bash
sky.exceptions.ClusterNotUpError: Tailing logs: skipped for cluster 'test-back-compat-buildkite-4' (status: INIT). It is only allowed for UP clusters. Wait for a launch to finish, or use this command to try to transition the cluster to UP: sky start test-back-compat-buildkite-4
```

Comment (Michaelvll) [reply sky/cli.py:1257]: We should not wait until it is not `INIT` as it is not guaranteed that the cluster will transfer to INIT.

We should do something similar as our `sky jobs logs`, i.e. not erroring out on `INIT` state, but, instead, try to tail the logs with best effort when it is `INIT` state.

Also, this is better to be separated out from the current PR.

File: tests/backward_compatibility_tests.sh [hunk 2]
```diff
@@ -10,13 +10,37 @@
 #   cd skypilot-repo
 #   git checkout <feature branch>
 #   pip uninstall -y skypilot ray
-#   bash tests/backward_compatibility_tests.sh
+#   bash tests/backward_compatibility_tests.sh [need_launch] [start_from] [base_branch]
```
Comment (Michaelvll) [inline tests/backward_compatibility_tests.sh:13]: Nice!

File: tests/backward_compatibility_tests.sh [hunk 3]
```diff
@@ -26,13 +50,107 @@ uuid=$(uuidgen)
...
+  set -vx  # Re-enable verbose tracing
+
+  # Show which sky is being used.
+  which sky
+
+  # Change to the environment directory
+  cd "$absolute_dir" || {
+    echo "Failed to change directory to $absolute_dir"
+    exit 1
+  }
```
Comment (Michaelvll) [inline tests/backward_compatibility_tests.sh:99]: Why do we need this? We only need to `cd` during the initial setup.

File: tests/backward_compatibility_tests.sh [hunk 4]
```diff
@@ -26,13 +50,107 @@ uuid=$(uuidgen)
...
+
+  # API operations: Restart the API if restart_api_server is 1
+  if [ "$restart_api_server" -eq 1 ]; then
+    sky api stop || true  # Stop, ignoring errors if it's not running
+    if ! sky api start; then
+      echo "Failed to start sky API"
+      exit 1
+    fi
+  fi
```
Comment (Michaelvll) [inline tests/backward_compatibility_tests.sh:108]: How about we just `sky api stop` whenever we `deactivate_env`, so that we don't need to handle the case whether we should restart the API server during the activate env

File: tests/backward_compatibility_tests.sh [hunk 5]
```diff
@@ -26,13 +50,107 @@ uuid=$(uuidgen)
...
+
+# Function to activate the base environment and change directory
+activate_base_env() {
+  activate_env "base" "${1:-1}"
+}
+
+# Function to activate the current environment and return to the initial directory
+activate_current_env() {
+  activate_env "current" "${1:-1}"
+}
```
Comment (Michaelvll) [inline tests/backward_compatibility_tests.sh:125]: These functions seems a bit shallow, we don't need them?

File: tests/backward_compatibility_tests.sh [hunk 6]
```diff
@@ -26,13 +50,107 @@ uuid=$(uuidgen)
...
+  echo "${status_output}" | grep UP || {
+    echo "Cluster ${cluster_name} failed to reach UP status"
+    echo "${status_output}"
+    exit 1
+  }
+}
+
+
+
+git clone -b ${base_branch} https://github.com/skypilot-org/skypilot.git $ABSOLUTE_BASE_VERSION_DIR || true
```
Comment (Michaelvll) [inline tests/backward_compatibility_tests.sh:147]: What if the dir exists? This will not clone the specified branch to the dir right? We should do checkout if the dir exists.

File: tests/backward_compatibility_tests.sh [hunk 7]
```diff
@@ -109,106 +222,97 @@ sky logs ${CLUSTER_NAME} 2
...
+activate_base_env
 sky launch --cloud ${CLOUD} -y --cpus 2 --num-nodes 2 -c ${CLUSTER_NAME}-2 examples/minimal.yaml
-deactivate
-source ~/sky-back-compat-current/bin/activate
-rm -r  ~/.sky/wheels || true
+deactivate_env
+activate_current_env
 sky stop -y ${CLUSTER_NAME}-2
 sky start -y ${CLUSTER_NAME}-2
+wait_until_cluster_up ${CLUSTER_NAME}-2
```
Comment (Michaelvll) [inline tests/backward_compatibility_tests.sh:236]: Why do we need to wait until UP? Our `sky start` should supposedly make sure that the cluster is in UP status once finished.

**Review by Michaelvll — COMMENTED** [review]
Thanks @zpoint! Please see the comments below

File: sky/cli.py (hunk 1 above)
Comment (zpoint) [reply sky/cli.py:1257]: issue created: 
#4753 
#4752

File: tests/backward_compatibility_tests.sh (hunk 3 above)
Comment (zpoint) [reply tests/backward_compatibility_tests.sh:99]: Because we have `PYTHONPATH=":$(pwd)"` when buildkite runs, this is the code that fixes the `backward compatibility test does not raise an error` Now, after the fix, it raises an error.

We change the `pwd` after modifying the environment to ensure Python loads from the correct codebase.

File: tests/backward_compatibility_tests.sh (hunk 5 above)
Comment (zpoint) [reply tests/backward_compatibility_tests.sh:125]: removed

File: tests/backward_compatibility_tests.sh (hunk 7 above)
Comment (zpoint) [reply tests/backward_compatibility_tests.sh:236]: see #4753

Removed here, will resolve in other PR.

File: tests/backward_compatibility_tests.sh (hunk 6 above)
Comment (zpoint) [reply tests/backward_compatibility_tests.sh:108]: This [line](https://github.com/skypilot-org/skypilot/blob/24d1584abfde43d625b0b4d4e8f8a0ae20161cbd/tests/backward_compatibility_tests.sh#L39) `rm -rf` ensures that directory not exists

File: tests/backward_compatibility_tests.sh (hunk 4 above)
Comment (zpoint) [reply tests/backward_compatibility_tests.sh]: This command will block:

```bash
sky status ${CLUSTER_NAME} | grep ${CLUSTER_NAME} | grep UP
```

If we let `sky status` launch the server, I think it didn't end the command properly due to the output buffer or pipe. I suggest keeping it as it is.

File: tests/backward_compatibility_tests.sh (hunk 6 above)
Comment (Michaelvll) [reply tests/backward_compatibility_tests.sh:118]: in that case why do we need this `|| true`?

File: tests/backward_compatibility_tests.sh (hunk 3 above)
Comment (Michaelvll) [reply tests/backward_compatibility_tests.sh:99]: why we need the `PYTHONPATH` to be set in the first place then? Can't we just rely on the `source ~/.../activate` to make sure the python is set to the correct env?

**Review by Michaelvll — COMMENTED** [review]
Thanks @zpoint! Left two quick questions

File: tests/backward_compatibility_tests.sh (hunk 6 above)
Comment (zpoint) [reply tests/backward_compatibility_tests.sh:118]: Removed

File: tests/backward_compatibility_tests.sh (hunk 3 above)
Comment (zpoint) [reply tests/backward_compatibility_tests.sh:99]: `PYTHONPATH` removed, I think it's a legacy env.

Even after removing `PYTHONPATH`, we still need the `cd` command. Test conclusion:

```
(sky) (sky) ➜  skypilot git:(dev/zeping/robust_backward_compatibility) source /Users/zepingguo/sky-back-compat-base/bin/activate
(sky-back-compat-base) (sky) (sky) ➜  skypilot git:(dev/zeping/robust_backward_compatibility) python -c 'import sky; print(sky.__version__)'
2.0.0-dev0
(sky-back-compat-base) (sky) (sky) ➜  skypilot git:(dev/zeping/robust_backward_compatibility) ✗ sky -v      
skypilot, version 1.0.0-dev0
(sky-back-compat-base) (sky) (sky) ➜  skypilot git:(dev/zeping/robust_backward_compatibility) ✗ 
```

* The `python` command imports from the current working directory by default.
* The `sky` command imports from the pip path by default.
* `sky api start` imports from the current working directory by default. After `sky start` (Might be because the subprocess changes the env.)

From the official [doc](https://docs.python.org/3/library/sys_path_init.html):

> The first entry in the module search path is the directory that contains the input script.

It completely depends on how the shell triggers the Python process. Everything happens implicitly. I think it's better to keep the `cd` command as it is to make it explicit.

Comment (zpoint) [reply tests/backward_compatibility_tests.sh:99]: We can trigger the expected [failure](https://buildkite.com/skypilot-1/quicktest-core/builds/144#01952c23-6f66-4194-b404-98302dd7fc04) after adding the `cd`.  
But it always [succeeds](https://buildkite.com/skypilot-1/quicktest-core/builds/143) when I remove the `cd`, even though I expect it to fail.

Comment (Michaelvll) [reply tests/backward_compatibility_tests.sh:99]: `python -c` imports the current directory is expected behavior, but `sky` CLI should supposedly invoke the python module in the pip list.

> The sky command imports from the pip path by default, which works before client-server infra but not after.

What do you mean by this? If that does not hold after client-server, this is a bug, and do you know the reason?

> The sky api start server imports from the current working directory by default. After sky start (likely because the subprocess changes the environment).

Could you elaborate? `sky api start` should use the `sky` module in the current python env. If not, it is a bug we should fix.

Comment (Michaelvll) [reply tests/backward_compatibility_tests.sh:99]: Also, the shell history you show above is fine, right? It is expected that `python -c` looks for the `sky` folder in the current dir first, but `sky` CLI should respect the current python env.

I don't see why this affect the backward compatibility test. Could you point me to where this cause the inconsistency for the test result?

Comment (zpoint) [reply tests/backward_compatibility_tests.sh:99]: > Also, the shell history you show above is fine, right? It is expected that python -c looks for the sky folder in the current dir first, but sky CLI should respect the current python env.

Yes thats right

Comment (zpoint) [reply tests/backward_compatibility_tests.sh:99]: > Could you elaborate?

I have two sky codebase, one in current pwd, set the version to `2.0.0-dev0`
One installed via pip under virtual env `sky-back-compat-base`, set the version to `2.0.0-dev0`


Log in client side:

```bash
# prints the version in virtual env
(sky-back-compat-base) (sky) (sky) ➜  skypilot git:(dev/zeping/robust_backward_compatibility) ✗ sky -v
skypilot, version 1.0.0-dev0
# prints the version in pwd
(sky-back-compat-base) (sky) (sky) ➜  skypilot git:(dev/zeping/robust_backward_compatibility) ✗ python -m sky.cli -v
skypilot, version 2.0.0-dev0
# I add some print code, it prints the version in virtual env in sky api start which is expected
(sky-back-compat-base) (sky) (sky) ➜  skypilot git:(dev/zeping/robust_backward_compatibility) ✗ sky api start
Failed to connect to SkyPilot API server at http://127.0.0.1:46580. Starting a local server.
sky.__version__: 1.0.0-dev0
/Users/zepingguo/sky-back-compat-base/bin/python3 -m sky.server.server --host 127.0.0.1 > /Users/zepingguo/.sky/api_server/server.log 2>&1 < /dev/null
✓ SkyPilot API server started.
├── SkyPilot API server: http://127.0.0.1:46580
└── View API server logs at: ~/.sky/api_server/server.log
(sky-back-compat-base) (sky) (sky) ➜  skypilot git:(dev/zeping/robust_backward_compatibility) ✗ 
```

Log in server side:

After the subprocess launches the server via `python -m`, api server loads the code from pwd instead of virtual env

```bash
I 02-22 21:17:36 httptools_impl.py:476] 127.0.0.1:63437 - "GET /api/health HTTP/1.1" 200
I 02-22 21:17:36 executor.py:310] Starting Worker(id=2, schedule_type=long) with pid 97772
I 02-22 21:17:36 executor.py:313] 2.0.0-dev0
I 02-22 21:17:36 executor.py:310] Starting Worker(id=1, schedule_type=short) with pid 97774
I 02-22 21:17:36 executor.py:313] 2.0.0-dev0
I 02-22 21:17:36 executor.py:310] Starting Worker(id=0, schedule_type=long) with pid 97770
I 02-22 21:17:36 executor.py:313] 2.0.0-dev0
I 02-22 21:17:36 executor.py:310] Starting Worker(id=1, schedule_type=long) with pid 97771
I 02-22 21:17:36 executor.py:313] 2.0.0-dev0
I 02-22 21:17:36 executor.py:310] Starting Worker(id=3, schedule_type=long) with pid 97773
I 02-22 21:17:36 executor.py:313] 2.0.0-dev0
```

^ @Michaelvll

Comment (Michaelvll) [reply tests/backward_compatibility_tests.sh:99]: That is a good investigation! In that case, we can always keep the current directory to be always in the parent directory, i.e. none of the codebase?

Also, we should raise an issue for the issue above. We should have the `sky api start` respect the module in python env

Comment (zpoint) [reply tests/backward_compatibility_tests.sh]: Issue created #4801

Comment (zpoint) [reply tests/backward_compatibility_tests.sh]: Also, add a `cd $ABSOLUTE_PARENT_DIR` in the activate env function. It should be good to go now.

Comment (Michaelvll) [reply tests/backward_compatibility_tests.sh]: Can we just have the `cd $ABSOLUTE_PARENT_DIR` at the beginning instead of in every activate env?

Comment (zpoint) [reply tests/backward_compatibility_tests.sh]: image
We need to change to the sky codebase directory for the test case to work. We can't keep the current directory in the parent folder.

**Review by Michaelvll — APPROVED** [review]
Thanks for the fix @zpoint! LGTM. This fixes the issue that the backward compatiblity fail to work with the master.

In future PRs, we still need to improve this backward compatiblity test to better catch backward compatibility issues, e.g. we need to have the tests in this file catching issues that not raised with non-zero returncode : )
